mongo/crud_examples_test.go

```diff
@@ -0,0 +1,525 @@
+	command := bson.D{{"create", "capped-coll"}, {"capped", true}, {"size", 4096}}
+	opts := options.RunCmd().SetReadPreference(readpref.Primary())
+	var result bson.M
+	if err := db.RunCommand(context.TODO(), command, opts).Decode(&result); err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println(result)
+}
+
+func ExampleDatabase_RunCommandCursor() {
+	var db *mongo.Database
+
+	// run a find command against collection "foo"
+	// specify the ReadPreference option to explicitly set the read preference to primary
+	command := bson.D{{"find", "foo"}}
+	opts := options.RunCmd().SetReadPreference(readpref.Primary())
+	cursor, err := db.RunCommandCursor(context.TODO(), command, opts)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// get a list of all returned documents and print them out
+	// see the mongo.Cursor documentation for more examples of using cursors
+	var results []bson.M
+	if err = cursor.All(context.TODO(), &results); err != nil {
+		log.Fatal(err)
+	}
+	for _, result := range results {
+		fmt.Println(result)
+	}
+}
+
+func ExampleDatabase_Watch() {
+	var db *mongo.Database
+
+	// specify a pipeline that will only match "insert" events
+	// specify the MaxAwaitTimeOption to have each attempt wait two seconds for new documents
+	matchStage := bson.D{{"$match", bson.D{{"operationType", "insert"}}}}
+	opts := options.ChangeStream().SetMaxAwaitTime(2 * time.Second)
+	changeStream, err := db.Watch(context.TODO(), mongo.Pipeline{matchStage}, opts)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// print out all change stream events in the order they're received
+	// see the mongo.ChangeStream documentation for more examples of using change streams
+	for changeStream.Next(context.TODO()) {
+		fmt.Println(changeStream.Current)
+	}
+}
+
+// Collection examples
+
+func ExampleCollection_Aggregate() {
+	var coll *mongo.Collection
+
+	// specify a pipeline that will return the number of times each name appears in the collection
+	// specify the MaxTime option to limit the amount of time the operation can run on the server
+	groupStage := bson.D{
+		{"$group", bson.D{
+			{"_id", "$name"},
+			{"numTimes", bson.D{
+				{"$sum", 1},
+			}},
+		}},
+	}
+	opts := options.Aggregate().SetMaxTime(2 * time.Second)
+	cursor, err := coll.Aggregate(context.TODO(), mongo.Pipeline{groupStage}, opts)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// get a list of all returned documents and print them out
+	// see the mongo.Cursor documentation for more examples of using cursors
+	var results []bson.M
+	if err = cursor.All(context.TODO(), &results); err != nil {
+		log.Fatal(err)
+	}
+	for _, result := range results {
+		fmt.Printf("name %v appears %v times\n", result["_id"], result["numTimes"])
+	}
+}
+
+func ExampleCollection_BulkWrite() {
+	var coll *mongo.Collection
+
+	// insert {name: "Alice"} and delete {name: "Bob"}
+	// set the Ordered option to false to allow both operations to happen even if one of them errors
+	models := []mongo.WriteModel{
```

Comment:
I suspect that by far the most common scenario for use of this method is a bulk upsert, so let's use that as an example instead.

Reply:
Can you explain what a bulk upsert is?

Reply:
Sure.  It's a bulk write, where each WriteModel is an UpdateOneModel with `upsert: true`.

Reply:
done.